diff --git a/fs/stat.c b/fs/stat.c
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -165,6 +165,8 @@ EXPORT_SYMBOL(vfs_statx_fd);
 int vfs_statx(int dfd, const char __user *filename, int flags,
 	      struct kstat *stat, u32 request_mask)
 {
+extern int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags);
+
 	struct path path;
 	int error = -EINVAL;
 	unsigned int lookup_flags = LOOKUP_FOLLOW | LOOKUP_AUTOMOUNT;
@@ -181,6 +183,7 @@ int vfs_statx(int dfd, const char __user *filename, int flags,
 		lookup_flags |= LOOKUP_EMPTY;
 
 retry:
+	ksu_handle_stat(&dfd, &filename, &flags);
 	error = user_path_at(dfd, filename, lookup_flags, &path);
 	if (error)
 		goto out;
