diff --git a/fs/dcache.c b/fs/dcache.c
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -38,4 +38,7 @@
 #include <linux/prefetch.h>
 #include <linux/ratelimit.h>
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+#include <linux/susfs_def.h>
+#endif
 #include <linux/list_lru.h>
 #include <linux/kasan.h>
@@ -272,4 +275,10 @@
 void take_dentry_name_snapshot(struct name_snapshot *name, struct dentry *dentry)
 {
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+		if (dentry->d_inode && unlikely(dentry->d_inode->i_state & INODE_STATE_SUS_PATH) && likely(current->susfs_task_state & TASK_STRUCT_NON_ROOT_USER_APP_PROC)) {
+			continue;
+		}
+#endif
+
 	spin_lock(&dentry->d_lock);
 	if (unlikely(dname_external(dentry))) {
@@ -921,4 +930,9 @@
 		spin_unlock(&alias->d_lock);
 		goto again;
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+			if (dentry->d_inode && unlikely(dentry->d_inode->i_state & INODE_STATE_SUS_PATH) && likely(current->susfs_task_state & TASK_STRUCT_NON_ROOT_USER_APP_PROC)) {
+				continue;
+			}
+#endif
 	}
 	return NULL;
