diff --git a/fs/proc/cmdline.c b/fs/proc/cmdline.c
--- a/fs/proc/cmdline.c
+++ b/fs/proc/cmdline.c
@@ -27,4 +27,14 @@
 #endif
 
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+extern int susfs_spoof_cmdline_or_bootconfig(struct seq_file *m);
+#endif
+
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+	if (!susfs_spoof_cmdline_or_bootconfig(m)) {
+		seq_putc(m, '\n');
+		return 0;
+	}
+#endif
 static int cmdline_proc_show(struct seq_file *m, void *v)
 {
