diff --git a/fs/proc/fd.c b/fs/proc/fd.c
--- a/fs/proc/fd.c
+++ b/fs/proc/fd.c
@@ -11,4 +11,7 @@
 #include <linux/fs.h>
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+#include <linux/susfs_def.h>
+#endif
 #include <linux/proc_fs.h>
 
@@ -21,4 +24,7 @@
 	struct files_struct *files = NULL;
 	int f_flags = 0, ret = -ENOENT;
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+	struct mount *mnt = NULL;
+#endif
 	struct file *file = NULL;
 	struct task_struct *task;
@@ -51,4 +57,15 @@
 
 	if (ret)
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+	mnt = real_mount(file->f_path.mnt);
+	if (likely(current->susfs_task_state & TASK_STRUCT_NON_ROOT_USER_APP_PROC) &&
+			mnt->mnt_id >= DEFAULT_SUS_MNT_ID) {
+		for (; mnt->mnt_id >= DEFAULT_SUS_MNT_ID; mnt = mnt->mnt_parent) { }
+	}
+	seq_printf(m, "pos:\t%lli\nflags:\t0%o\nmnt_id:\t%i\n",
+			(long long)file->f_pos, f_flags,
+			mnt->mnt_id);
+#else
+#endif
 		return ret;
 
